File: src/fridaymario/listeners/Compilation.java
package fridaymario.listeners;

import com.intellij.openapi.Disposable;
import com.intellij.openapi.project.Project;

public abstract class Compilation implements Restartable {
	@Override public void start(Disposable disposable) {
	}

	public static Factory factory = (project, listener) -> new Compilation() {
	};

	public interface Factory {
		Compilation create(Project project, Listener listener);
	}

	public interface Listener {
		void compilationSucceeded();

		void compilationFailed();
	}
}

File: src/fridaymario/listeners/Compilation.kt
package fridaymario.listeners

import com.intellij.openapi.Disposable
import com.intellij.openapi.project.Project

abstract class Compilation: Restartable {
    override fun start(disposable: Disposable) {}

    interface Factory {
        fun create(project: Project, listener: Listener): Compilation
    }

    interface Listener {
        fun compilationSucceeded()
        fun compilationFailed()
    }

    companion object {
        var factory: Factory = object: Factory {
            override fun create(project: Project, listener: Listener) = object: Compilation() {}
        }
    }
}